fl4health/clients/ditto_mkmmd_client.py

```diff
@@ -0,0 +1,483 @@
+        """
+        The parameters being pass are to be routed to the global model and saved as the initial global model tensors to
+        be used in a penalty term in training the local model. In the first fitting round, we assume the both the
+        global and local models are being initialized and use the FullParameterExchanger() to set all model weights.
+        Args:
+            parameters (NDArrays): Parameters have information about model state to be added to the relevant client
+                model (global model for all but the first step of Ditto)
+            config (Config): The config is sent by the FL server to allow for customization in the function if desired.
+            fitting_round (bool): Boolean that indicates whether the current federated learning
+                round is a fitting round or an evaluation round.
+                This is used to help determine which parameter exchange should be used for pulling parameters.
+                A full parameter exchanger is only used if the current federated learning round is the very
+                first fitting round.
+        """
+        # Make sure that the proper components exist.
+        assert self.global_model is not None and self.model is not None
+        assert self.parameter_exchanger is not None and isinstance(self.parameter_exchanger, FullParameterExchanger)
+
+        current_server_round = self.narrow_config_type(config, "current_server_round", int)
+        if current_server_round == 1 and fitting_round:
+            log(INFO, "Initializing the global and local models weights for the first time")
+            self.initialize_all_model_weights(parameters, config)
+        else:
+            # Route the parameters to the GLOBAL model in Ditto
+            assert self.parameter_exchanger is not None
+            log(INFO, "Setting the global model weights")
+            self.parameter_exchanger.pull_parameters(parameters, self.global_model, config)
+
+    def initialize_all_model_weights(self, parameters: NDArrays, config: Config) -> None:
+        """
+        If this is the first time we're initializing the model weights, we initialize both the global and the local
+        weights together.
+
+        Args:
+            parameters (NDArrays): Model parameters to be injected into the client model
+            config (Config): The config is sent by the FL server to allow for customization in the function if desired.
+        """
+        self.parameter_exchanger.pull_parameters(parameters, self.model, config)
+        self.parameter_exchanger.pull_parameters(parameters, self.global_model, config)
+
+    def update_before_train(self, current_server_round: int) -> None:
+        assert isinstance(self.model, nn.Module)
+        # Clone and freeze the initial weights GLOBAL MODEL. These are used to form the Ditto local
+        # update penalty term.
+        self.init_global_model = self.clone_and_freeze_model(self.global_model)
```

Comment:
I like doing this here, but perhaps instead of cloning the model here, we can just clone and detach the weights to be stored as in the original Ditto implementation like this 
```
# Saving the initial weights GLOBAL MODEL weights and detaching them so that we don't compute gradients with
# respect to the tensors. These are used to form the Ditto local update penalty term.
self.initial_global_tensors = [
    initial_layer_weights.detach().clone() for initial_layer_weights in self.global_model.parameters()
]
```

Reply:
Below, I see why we clone the initial global model here. So we should definitely leave that, but I think it's still a good idea to save the initial tensors globally, rather than forming them at every step since they are constant.

Reply:
In new update I basically extend previous ditto_client. so now I have both tensors and init global model in ditto mlmmd client

Reply:
Sounds good!